packages/kmod-5.15-nvidia/kmod-5.15-nvidia.spec

```diff
@@ -0,0 +1,324 @@
+# libraries into the containers
+sed -e 's|__LIBDIR__|%{_cross_libdir}|' %{S:303} | sed -e 's|__NVIDIA_VERSION__|%{tesla_515}|' \
+  > nvidia-tesla-%{tesla_515}.conf
+install -m 0644 nvidia-tesla-%{tesla_515}.conf %{buildroot}%{_cross_factorydir}%{_cross_sysconfdir}/ld.so.conf.d/
+
+# driver
+install kernel/nvidia.mod.o %{buildroot}%{_cross_datadir}/nvidia/tesla/%{tesla_515}/module-objects.d
+install kernel/nvidia/nv-interface.o %{buildroot}%{_cross_datadir}/nvidia/tesla/%{tesla_515}/module-objects.d
+install kernel/nvidia/nv-kernel.o_binary %{buildroot}%{_cross_datadir}/nvidia/tesla/%{tesla_515}/module-objects.d/nv-kernel.o
+
+# uvm
+install kernel/nvidia-uvm.mod.o %{buildroot}%{_cross_datadir}/nvidia/tesla/%{tesla_515}/module-objects.d
+install kernel/nvidia-uvm.o %{buildroot}%{_cross_datadir}/nvidia/tesla/%{tesla_515}/module-objects.d
+
+# modeset
+install kernel/nvidia-modeset.mod.o %{buildroot}%{_cross_datadir}/nvidia/tesla/%{tesla_515}/module-objects.d
+install kernel/nvidia-modeset/nv-modeset-interface.o %{buildroot}%{_cross_datadir}/nvidia/tesla/%{tesla_515}/module-objects.d
+install kernel/nvidia-modeset/nv-modeset-kernel.o %{buildroot}%{_cross_datadir}/nvidia/tesla/%{tesla_515}/module-objects.d
+
+# peermem
+install kernel/nvidia-peermem.mod.o %{buildroot}%{_cross_datadir}/nvidia/tesla/%{tesla_515}/module-objects.d
+install kernel/nvidia-peermem/nvidia-peermem.o %{buildroot}%{_cross_datadir}/nvidia/tesla/%{tesla_515}/module-objects.d
+
+# drm
+install kernel/nvidia-drm.mod.o %{buildroot}/%{_cross_datadir}/nvidia/tesla/%{tesla_515}/module-objects.d
+install kernel/nvidia-drm.o %{buildroot}/%{_cross_datadir}/nvidia/tesla/%{tesla_515}/module-objects.d
+
+# Binaries
+install -m 755 nvidia-smi %{buildroot}%{_cross_libexecdir}/nvidia/tesla/bin/%{tesla_515}
+install -m 755 nvidia-debugdump %{buildroot}%{_cross_libexecdir}/nvidia/tesla/bin/%{tesla_515}
+install -m 755 nvidia-cuda-mps-control %{buildroot}%{_cross_libexecdir}/nvidia/tesla/bin/%{tesla_515}
+install -m 755 nvidia-cuda-mps-server %{buildroot}%{_cross_libexecdir}/nvidia/tesla/bin/%{tesla_515}
+%if "%{_cross_arch}" == "x86_64"
+install -m 755 nvidia-ngx-updater %{buildroot}%{_cross_libexecdir}/nvidia/tesla/bin/%{tesla_515}
+%endif
+
+# We install all the libraries, and filter them out in the 'files' section, so we can catch
+# when new libraries are added
+install -m 755 *.so* %{buildroot}/%{tesla_515_libdir}/
+
+# This library has the same SONAME as libEGL.so.1.1.0, this will cause collisions while
+# the symlinks are created. For now, we only symlink libEGL.so.1.1.0.
+EXCLUDED_LIBS="libEGL.so.%{tesla_515}"
+
+for lib in $(find . -maxdepth 1 -type f -name 'lib*.so.*' -printf '%%P\n'); do
+  [[ "${EXCLUDED_LIBS}" =~ "${lib}" ]] && continue
+  soname="$(%{_cross_target}-readelf -d "${lib}" | awk '/SONAME/{print $5}' | tr -d '[]')"
+  [ -n "${soname}" ] || continue
+  [ "${lib}" == "${soname}" ] && continue
+  [ -e %{buildroot}/%{tesla_515_libdir}/"${soname}" ] && continue
```

Comment:
There are a few libraries in the binary for which the soname'd file is provided in the archive:

```
libnvidia-gtk3.so.515.65.01
libnvidia-eglcore.so.515.65.01
libnvidia-rtcore.so.515.65.01
libGLX.so.0
libnvidia-glvkspirv.so.515.65.01
libnvidia-tls.so.515.65.01
libnvidia-gtk2.so.515.65.01
libGLdispatch.so.0
libnvidia-wayland-client.so.515.65.01
libnvidia-compiler.so.515.65.01
libnvidia-glsi.so.515.65.01
libnvidia-glcore.so.515.65.01
libOpenGL.so.0
```

However, there are a few that still require the symlink:

```
libGLESv2_nvidia.so.515.65.01
libnvoptix.so.515.65.01
libnvidia-egl-wayland.so.1.1.9
libGL.so.1.7.0
libnvidia-allocator.so.515.65.01
libvdpau_nvidia.so.515.65.01
libnvidia-ngx.so.515.65.01
libEGL.so.1.1.0
libnvidia-nvvm.so.515.65.01
libnvidia-encode.so.515.65.01
libGLX_nvidia.so.515.65.01
libGLESv2.so.2.1.0
libnvidia-egl-gbm.so.1.1.0
libEGL.so.515.65.01
libOpenCL.so.1.0.0
libnvidia-fbc.so.515.65.01
libnvidia-opticalflow.so.515.65.01
libGLESv1_CM.so.1.2.0
libGLESv1_CM_nvidia.so.515.65.01
libnvidia-cfg.so.515.65.01
libnvidia-ptxjitcompiler.so.515.65.01
libnvidia-opencl.so.515.65.01
libnvidia-ml.so.515.65.01
libcuda.so.515.65.01
libEGL_nvidia.so.515.65.01
libnvcuvid.so.515.65.01
```

Reply:
> There are a few libraries in the binary for which the soname'd file is provided in the archive

The "file" or the "symlink"? If we test (via `[ -L ${link} ]`) that it is already a link then OK. If it is a regular file with the same name as the library, then that is not really OK, because it creates ambiguity as to which one the dynamic loader will select.

I would tend to prefer testing that it's a link and then removing it and recreating our own. If it exists and it's not a link, do something else - diff it against the target and remove it if they're the same, then create our link. Otherwise it's an exceptional case and I need more details to advise.

Reply:
I made a mistake while I was evaluating the symlinks that existed for each library (fish-shell bit me :anger: ). Anyways, I confirmed that neither the symlinks nor the files exist in the NVIDIA run archives for these libraries

```Missing 'libGLESv2_nvidia.so.2' for 'libGLESv2_nvidia.so.515.65.01'
Missing 'libnvoptix.so.1' for 'libnvoptix.so.515.65.01'
Missing 'libnvidia-egl-wayland.so.1' for 'libnvidia-egl-wayland.so.1.1.9'
Missing 'libGL.so.1' for 'libGL.so.1.7.0'
Missing 'libnvidia-allocator.so.1' for 'libnvidia-allocator.so.515.65.01'
Missing 'libvdpau_nvidia.so.1' for 'libvdpau_nvidia.so.515.65.01'
Missing 'libnvidia-ngx.so.1' for 'libnvidia-ngx.so.515.65.01'
Missing 'libEGL.so.1' for 'libEGL.so.1.1.0'
Missing 'libnvidia-nvvm.so.4' for 'libnvidia-nvvm.so.515.65.01'
Missing 'libnvidia-encode.so.1' for 'libnvidia-encode.so.515.65.01'
Missing 'libGLX_nvidia.so.0' for 'libGLX_nvidia.so.515.65.01'
Missing 'libGLESv2.so.2' for 'libGLESv2.so.2.1.0'
Missing 'libnvidia-egl-gbm.so.1' for 'libnvidia-egl-gbm.so.1.1.0'
Missing 'libEGL.so.1' for 'libEGL.so.515.65.01'
Missing 'libOpenCL.so.1' for 'libOpenCL.so.1.0.0'
Missing 'libnvidia-fbc.so.1' for 'libnvidia-fbc.so.515.65.01'
Missing 'libnvidia-opticalflow.so.1' for 'libnvidia-opticalflow.so.515.65.01'
Missing 'libGLESv1_CM.so.1' for 'libGLESv1_CM.so.1.2.0'
Missing 'libGLESv1_CM_nvidia.so.1' for 'libGLESv1_CM_nvidia.so.515.65.01'
Missing 'libnvidia-cfg.so.1' for 'libnvidia-cfg.so.515.65.01'
Missing 'libnvidia-ptxjitcompiler.so.1' for 'libnvidia-ptxjitcompiler.so.515.65.01'
Missing 'libnvidia-opencl.so.1' for 'libnvidia-opencl.so.515.65.01'
Missing 'libnvidia-ml.so.1' for 'libnvidia-ml.so.515.65.01'
Missing 'libcuda.so.1' for 'libcuda.so.515.65.01'
Missing 'libEGL_nvidia.so.0' for 'libEGL_nvidia.so.515.65.01'
Missing 'libnvcuvid.so.1' for 'libnvcuvid.so.515.65.01'
```

This is the script that I used to verify which libraries are missing their `SONAME` symlink, and which don't need it:

```bash
#! /usr/bin/env bash

for lib in $(find . -maxdepth 1 -type f -name 'lib*.so.*' -printf '%P\n'); do
  soname="$(readelf -d "${lib}" | awk '/SONAME/{print $5}' | tr -d '[]')"
  [ -n "${soname}" ] || continue
  [ "${lib}" == "${soname}" ] && continue
  [ ! -e "${soname}" ] && echo "Missing '${soname}' for '${lib}'"
done

```